docs/guides/streaming.md

```diff
@@ -0,0 +1,369 @@
+
+Remix takes advantage of React 18's streaming and server-side support for `<Suspense />` boundaries using the [`deferred` Response][deferred-response] utility and [`<Deferred />`][deferred] component / [`useDeferredData`][usedeferreddata] hook. By using these APIs, you can solve both of these problems:
+
+1. You're data is no longer on a waterfall: document & data (in parallel) -> JavaScript
+2. Your can easily switch between streaming and waiting for the data
+
+Let's take a dive into how to accomplish this.
+
+### Enable React 18 Streaming
+
+First, to enable streaming with React 18, you'll update your `entry.server.tsx` file to use `renderToPipeableStream`. Here's a simple (and incomplete) version of that:
+
+```tsx filename=app/entry.server.tsx lines=[1-2,17,24,29,34]
+import { PassThrough } from "stream";
+import { renderToPipeableStream } from "react-dom/server";
+import { RemixServer } from "@remix-run/react";
+import { Response } from "@remix-run/node";
+import type {
+  EntryContext,
+  Headers,
+} from "@remix-run/node";
+
+export default function handleRequest(
+  request: Request,
+  responseStatusCode: number,
+  responseHeaders: Headers,
+  remixContext: EntryContext
+) {
+  return new Promise((resolve) => {
+    const { pipe } = renderToPipeableStream(
+      <RemixServer
+        context={remixContext}
+        url={request.url}
+      />,
+      {
+        onShellReady() {
+          const body = new PassThrough();
+
+          responseHeaders.set("Content-Type", "text/html");
+
+          resolve(
+            new Response(body, {
+              status: responseStatusCode,
+              headers: responseHeaders,
+            })
+          );
+          pipe(body);
+        },
+      }
+    );
+  });
+}
+```
+
+<details>
+  <summary>For a more complete example, expand this</summary>
+
+This handles errors and properly disables streaming for bots which you typically want to force waiting so you can display all the content for SEO purposes.
+
+```tsx filename=app/entry.server.tsx
+import { PassThrough } from "stream";
+import { renderToPipeableStream } from "react-dom/server";
+import { RemixServer } from "@remix-run/react";
+import { Response } from "@remix-run/node";
+import type {
+  EntryContext,
+  Headers,
+} from "@remix-run/node";
+import isbot from "isbot";
+
+const ABORT_DELAY = 5000;
+
+export default function handleRequest(
+  request: Request,
+  responseStatusCode: number,
+  responseHeaders: Headers,
+  remixContext: EntryContext
+) {
+  const callbackName = isbot(
+    request.headers.get("user-agent")
+  )
+    ? "onAllReady"
+    : "onShellReady";
+
+  return new Promise((resolve, reject) => {
+    let didError = false;
+
+    const { pipe, abort } = renderToPipeableStream(
+      <RemixServer
+        context={remixContext}
+        url={request.url}
+      />,
+      {
+        [callbackName]() {
+          let body = new PassThrough();
+
+          responseHeaders.set("Content-Type", "text/html");
```

Comment:
Lighthouse audits say that you should add the `charset` in the first 1024 bytes streamed from the server. This can also be done via a meta tag but I feel like this is the easiest place to add it. 

```suggestion
          responseHeaders.set("Content-Type", "text/html; charset=UTF-8");
```

Reply:
Perhaps we should update all the example to do this 🤔

Reply:
According to @mathiasbynens, it's a best practice to use the `<meta charset="utf-8">` (see https://github.com/GoogleChrome/lighthouse/issues/10023#issuecomment-558521860)

Reply:
Ok. I can see the argument to keep things as is. 

I using a CCS-in-JS library (I know it’s something the Remix community thinks should die) which automatically injects style tags into the top of the head in document. In my case, this bumped the meta tag to be too low in the document to work. Again, I understand that is an edge case so I will close this.